zcash_client_sqlite/src/wallet/init/migrations/full_account_ids.rs

```diff
@@ -0,0 +1,381 @@
+use std::{collections::HashSet, rc::Rc};
+
+use rusqlite::{params, Transaction};
+use schemer_rusqlite::RusqliteMigration;
+use secrecy::SecretVec;
+use uuid::Uuid;
+use zcash_keys::keys::HdSeedFingerprint;
+
+use crate::wallet::init::WalletMigrationError;
+
+use super::{add_account_birthdays, receiving_key_scopes, v_transactions_note_uniqueness};
+
+/// The migration that switched from presumed seed-derived account IDs to supporting
+/// HD accounts and all sorts of imported keys.
+pub(super) const MIGRATION_ID: Uuid = Uuid::from_u128(0x1b104345_f27e_42da_a9e3_1de22694da43);
+
+pub(crate) struct Migration {
+    pub(super) seed: Rc<Option<SecretVec<u8>>>,
+}
+
+impl schemer::Migration for Migration {
+    fn id(&self) -> Uuid {
+        MIGRATION_ID
+    }
+
+    fn dependencies(&self) -> HashSet<Uuid> {
+        [
+            receiving_key_scopes::MIGRATION_ID,
+            add_account_birthdays::MIGRATION_ID,
+            v_transactions_note_uniqueness::MIGRATION_ID,
+        ]
+        .into_iter()
+        .collect()
+    }
+
+    fn description(&self) -> &'static str {
+        "Replaces the `account` column in the `accounts` table with columns to support all kinds of account and key types."
+    }
+}
+
+impl RusqliteMigration for Migration {
+    type Error = WalletMigrationError;
+
+    fn up(&self, transaction: &Transaction) -> Result<(), WalletMigrationError> {
+        transaction.execute_batch(
+            r#"
+            PRAGMA foreign_keys = OFF;
+            PRAGMA legacy_alter_table = ON;
+
+            CREATE TABLE accounts_new (
+                id INTEGER NOT NULL PRIMARY KEY AUTOINCREMENT,
+                account_type INTEGER NOT NULL DEFAULT 0,
+                hd_seed_fingerprint BLOB,
+                hd_account_index INTEGER,
+                uvk TEXT NOT NULL,
+                birthday_height INTEGER NOT NULL,
+                recover_until_height INTEGER,
+                CHECK (
+                    (account_type = 0 AND hd_seed_fingerprint IS NOT NULL AND hd_account_index IS NOT NULL)
+                    OR
+                    (account_type = 1 AND hd_seed_fingerprint IS NULL AND hd_account_index IS NULL)
+                )
+            );
+            CREATE UNIQUE INDEX accounts_uvk ON accounts_new ("uvk");
+            "#,
+        )?;
+
+        // We require the seed *if* there are existing accounts in the table.
+        if transaction.query_row("SELECT COUNT(*) FROM accounts", [], |row| {
+            Ok(row.get::<_, u32>(0)? > 0)
+        })? {
+            if let Some(seed) = &self.seed.as_ref() {
+                let seed_id = HdSeedFingerprint::from_seed(seed);
```

Comment:
For safety, we should check that every account in the existing wallet is derived from the given seed, and return `WalletMigrationError::CorruptedData` if not (like we did in the `ufvk_support` migration).